cmd/healthcheck/main_test.go

```diff
@@ -0,0 +1,130 @@
+package main
+
+import (
+	"github.com/stretchr/testify/assert"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+)
+
+func TestHealthStatusHealthy(t *testing.T) {
+	if os.Getenv("FLAG") == "1" {
+		healthCheckDefaultUrl := "127.0.0.1:13133"
+		l, _ := net.Listen("tcp", healthCheckDefaultUrl)
+		server := httptest.NewUnstartedServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			rw.WriteHeader(http.StatusOK)
+		}))
+		server.Listener.Close()
+		server.Listener = l
+		server.Start()
+		os.Args = []string{"-port=13133"}
+
+		defer server.Close()
+		main()
+		return
+	} // Run the test in a subprocess
+	cmd := exec.Command(os.Args[0], "-test.run=TestHealthStatusHealthy") //nolint:gosec
+	cmd.Env = append(os.Environ(), "FLAG=1")
+
+	stdout, _ := cmd.StderrPipe()
+	er := cmd.Start()
+	if er != nil {
+		t.Fatal("Unable to start the testing sub-process.")
+	}
+
+	gotBytes, _ := ioutil.ReadAll(stdout)
+	got := string(gotBytes)
+	expectedErrorString := "STATUS: 200"
+
+	if !strings.Contains(got[:len(got)-1], expectedErrorString) {
+		t.Fatalf("Unexpected log message. Got %s but should contain %s", got[:len(got)-1], expectedErrorString)
+	}
+	err := cmd.Wait()
+	e, _ := err.(*exec.ExitError)
+	assert.Nil(t, e)
+
+}
+
+func TestHealthStatusUnhealthy(t *testing.T) {
+	if os.Getenv("FLAG") == "1" {
+		healthCheckDefaultUrl := "127.0.0.1:13133"
+		l, _ := net.Listen("tcp", healthCheckDefaultUrl)
+		server := httptest.NewUnstartedServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			rw.WriteHeader(http.StatusInternalServerError)
+		}))
+		server.Listener.Close()
+		server.Listener = l
+		server.Start()
+		os.Args = []string{"-port=13133"}
+
+		defer server.Close()
+		main()
+		return
+	} // Run the test in a subprocess
+	cmd := exec.Command(os.Args[0], "-test.run=TestHealthStatusUnhealthy") //nolint:gosec
```

Comment:
have we considered creating a function that is called by main that will return an error and a message. then you can test just that function and you don't need to create a subprocess. 

Alternatively, is there a way to call the binary of the application instead of resorting to this flag based mechanism?

Reply:
SubProcess is required as main exits (either with code 1 or 0). So even if I take that in separate func, then a subprocess would be required to test that function. Using binary for unit-testing may require further alignment/approval with team.

Reply:
I meant that subprocess can be replaced by a function and you could test just that function:

```go

	resp, err := http.Get(fmt.Sprint("http://", host, ":", *port, path))
	if err != nil {
		log.Fatalf("Unable to retrieve health status: %s", err.Error())
	}
	if resp != nil && resp.StatusCode == 200 {
		log.Printf("STATUS: %d", resp.StatusCode)
	} else {
		log.Fatalf("STATUS: %d", resp.StatusCode)
	}
```

transformed into:
```go
func executeHealthCheck(host, port, path) (String, error)) {
	resp, err := http.Get(fmt.Sprint("http://", host, ":", *port, path))

	if err != nil {
		return nill, fmt.Errorf("Unable to retrieve health status: %s", err.Error())
	}

	if resp.StatusCode != 200 {
		return nill, fmt.Errorf("STATUS: %d", resp.StatusCode)
	}

        return fmt.Sprint("STATUS: %d", resp.StatusCode), nill
}

func main() {
	const host = "127.0.0.1" // default host
	usedPort := "13133"      // default port
	const path = "/"         //default path
	generateCmd := flag.NewFlagSet("generate", flag.ExitOnError)
	port := generateCmd.String("port", usedPort, "Specify collector health-check port")

	if len(os.Args) > 1 {
		err := generateCmd.Parse(os.Args[1:])
		if err != nil {
			log.Fatalf("%s", err)
		}
	}

	validationErr := validatePort(*port)
	if validationErr != nil {
		log.Fatalf("%s", validationErr)
	}
       
        status, healthCheckError := executeHealthCheck(host, port, path)

        if healthCheckError {
                log.Fatalf(healthCheckError.Error())
        }

        log.Printf(status)
}
```

Then you can test just `executeHealthCheck` and with that you get rid of the complicated test setup. The idea is that you don't need to test that Fatalf exists the application with status code `1`.

Reply:
Thanks updated.